### PR TITLE
AppID/TBF: Add `ShortId` header

### PR DESCRIPTION
**Pull request by bradjc**
### Pull Request Overview

Tock's process loading mechanism can assign AppIds and ShortIds to processes. Since there are only ~2^32 ShortIds, Tock kernels need some (at least partially manual) method to assign ShortIds to avoid the issue where two different applications are assigned the same ShortId.

Right now we really only have one upstream method: using a CRC of the process's name to create a ShortId.

This would add a second: allowing the developer to set the ShortId in the TBF header of the process. An implementation of the `Compress` trait can then use this header to set the ShortId (and this PR includes an example of that). This is the successor to a version that was in the original AppId PR where the id would be stored in the footer. That was removed because the footer is not included in the integrity region. This moves it to the header which is checked.

A kernel that is concerned about trusted apps should be careful to only use this header if the app was correctly signed by a trusted party.

The header itself is simple and is just a u32 with the fixed ShortId.

This PR also includes a config board to help with testing this feature.

I have corresponding changes to the book and Tockloader.

### Testing Strategy

Running the config board and displaying the ShortId as set by the TBF header.

```
❯ tockloader inspect-tab
[INFO   ] No TABs passed to tockloader.
[STATUS ] Searching for TABs in subdirectories.
[INFO   ] Using: ['./build/blink.tab']
[STATUS ] Inspecting TABs...
TAB: blink
  build-date: 2024-06-06 18:58:13+00:00
  minimum-tock-kernel-version: 2.0
  tab-version: 1
  included architectures: cortex-m0, cortex-m3, cortex-m4, cortex-m7, rv32imac, rv32imc

 Which TBF to inspect further? cortex-m4

cortex-m4:
  version               : 2                       [0x0  ]
  header_size           :         76         0x4c
  total_size            :       8192       0x2000
  checksum              :              0x6e3c4860
  flags                 :          1          0x1
    enabled             : Yes
    sticky              : No
  TLV: Main (1)                                   [0x10 ]
    init_fn_offset      :         49         0x31
    protected_size      :          8          0x8
    minimum_ram_size    :       4916       0x1334
  TLV: Program (9)                                [0x20 ]
    init_fn_offset      :         49         0x31
    protected_size      :          8          0x8
    minimum_ram_size    :       4916       0x1334
    binary_end_offset   :       1612        0x64c
    app_version         :          0          0x0
  TLV: Package Name (3)                           [0x38 ]
    package_name        : blink
  TLV: ShortID (10)                               [0x44 ]
    short_id            : 551

TBF Footers
  Footer
    footer_size         :       6580       0x19b4
  Footer TLV: Credentials (128)                   [0x0  ]
    Type: Reserved (0)
    Length: 6572
```

### TODO or Help Wanted

n/a

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: libraries/tock-tbf/src/types.rs [hunk 1]
```diff
@@ -130,6 +130,7 @@ pub enum TbfHeaderTypes {
     TbfHeaderStoragePermissions = 7,
     TbfHeaderKernelVersion = 8,
     TbfHeaderProgram = 9,
+    TbfHeaderShortId = 10,
```
Comment (alistair23) [inline libraries/tock-tbf/src/types.rs:133]: The book should be updated as well

**Review by alistair23 — CHANGES_REQUESTED** [review]

File: libraries/tock-tbf/src/types.rs (hunk 1 above)
Comment (bradjc) [reply libraries/tock-tbf/src/types.rs:133]: https://github.com/tock/book/pull/40 ?

Comment (alistair23) [reply libraries/tock-tbf/src/types.rs:133]: Ah, it just wasn't linked in the PR description

**Review by ppannuto — APPROVED** [review]

**Review by alevy — APPROVED** [review]

File: capsules/system/src/process_checker/tbf.rs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        true
+    }
+
+    fn different_identifier_processes(
+        &self,
+        _process_a: &dyn Process,
+        _process_b: &dyn Process,
+    ) -> bool {
+        true
+    }
```
Comment (alistair23) [inline capsules/system/src/process_checker/tbf.rs:39]: Doesn't this break the unique guarantee? What if two applications have the same shortID in the header?

Comment (bradjc) [reply capsules/system/src/process_checker/tbf.rs:39]: That would mean they are the same application and only one will run. This check is for AppId, the ShortId check is done in the kernel checker.

**Pull request merged**